### DISCUSSION TITLE
LlamaParse Roadmap

### DISCUSSION THREAD

**Discussion by jerryjliu** [run-llama/llama_cloud_services#19] [Announcements]
Currently LlamaParse supports complex PDF documents as input. We will extend LlamaParse in the coming weeks / months to support the following:

- More file formats, starting with .docx and .pptx coming this month.
- Ability to retrieve image embeded in document coming this month.
- More output format better suited for Retrieval augmented generation than Markdown.
- Better support for complexe document layouts, including graphs and pie charts.
- Vertical specific document parser starting with Financial documents, invoices, forms and contract.

Of course, we will also work hard to address any issue that may arise. Please drop in the Issues tab!

**Comment by jamesqiyucai** [comment]
When will LlamaParse support Chinese?

**Reply by hexapode** [reply]
LlamaParse already support chinese. You should specify language=ch_sim or language= ch_trad in the api parameters. Once https://github.com/run-llama/llama_parse/pull/37 is merged, you should be able to do it with the python pluggin.

**Comment by DMAgit** [comment]
Is the source code of the parsing logic itself available, or will it be made available? As far as I can see, currently only the code which accesses the remote point is available, but not any of the parsing logic.

**Comment by relsas** [comment]
**Testing on Academic Papers**

Hi, I very much appreciate the initiative on LlamaParse as pdf-parsing is indeed a challenge. Putting this as an API is also cool and solves many installation issues. The idea of using LLM instructions to steer the parser is great, as is the separation of objects for tables and figures.
However, I've now spent quite some time on testing to parse academic papers, with mediocre success only. In particular, the following issues occur:
- A new page in the pdf always breaks the current paragraph, such that often sentences are interrupted and continued in the next paragraph.
- If the pdf has footer or header information, this is repeated over an over, sometimes even classified as headers. Such footer info is then put between the body text when the page switch occurs, so that one cannot reconstruct interrupted sentences/paragraphs (see above)    
- Footnotes are not systematically recognized, numeration is in most cases lost, and sometimes footnotes appear in the body text 
- Formulas are barely translated to Latex.

In most of these aspects, parsing by Nougat works much better (though this has problems on its own...e.g. with references or left out pages).
Any ideas (or maybe plans) how to improve  model performance on such complex types of pdfs?
Kind regards

**Comment by gittar** [comment]
Will there be LlamaParse on prem, i.e. a version which can be run locally without API calls? This would make its use für confidential documents possible. Also a "No"-answer would be helpful for decision making.

**Reply by martincollado** [reply]
I think that the "No" answer is an essential aspect. I'm trying to see on forums, google searches, etc. what are the decisions that you made on your platform about the data privacy in this product and is difficult to view a clear answer. 

Could you provide more information about it @hexapode ? I don't think that any company out there will throw private and sensitive information on this API before being aware of what's happening with their data.

**Reply by jordanparker6** [reply]
@hexapode would love to know if this feature is roadmapped. Would help to understand potential licensing costs if this were to be part of an on-prem stack for enterprise.

**Comment by guillaume-millot** [comment]
I have tried Llama Parse to parse complex financial documents. I found it to be a great and innovative tool.
However, often, it makes basic mistakes such as completely ignoring a big portion of the PDF.
It'd really be the best tool if it were more robust. Curious to know how much you're investing on llama parse core improvements vs other things.
Thank you!

**Comment by LDD19** [comment]
> Better support for complex document layouts, including graphs and pie charts.

Any plans to integrate this in the near future? Especially with GPT-4o being multimodal, filling the missing data in parsed charts/images shouldn't be an issue.

**Comment by aifirstd3v** [comment]
My PDF only has images :) I hope LlamaParse can read image :)

**Comment by kinthaiofficial** [comment]
A few things that would be high-value additions to LlamaParse from an agent pipeline perspective:

**Structured extraction with schema** — beyond OCR and layout parsing, the ability to extract into a specified Pydantic/JSON Schema directly. "Parse this invoice and return a `InvoiceSchema` object" rather than "return markdown, then I'll extract." Reduces one LLM hop in the pipeline.

**Incremental / delta parsing** — when a document is updated (new page added, table modified), parse only the changed sections and return a delta. Avoids re-parsing stable pages. Crucial for long-lived knowledge bases where documents evolve.

**Parse quality confidence** — a confidence score per extracted element (this table was cleanly parsed: 0.97; this blurry scan section: 0.61). Agents can use this to decide whether to re-fetch the source, flag for human review, or proceed with reduced confidence.

**Extraction lineage** — for agent systems, knowing "this fact came from page 4, table 2, row 3 of document X (version 2026-01-15)" is essential for citation and audit. Parse results should carry enough metadata for the agent to build a provenance trail.

**Batch prioritization** — when an agent submits 50 documents to parse, it should be able to flag 5 as urgent (needed for the current task) and 45 as background. Priority queue semantics rather than FIFO.

We've been building document-aware agents in KinthAI that need reliable parsing as input to their knowledge consolidation step: https://blog.kinthai.ai/why-character-ai-forgets-you-persistent-memory-architecture covers how parsing quality affects downstream memory reliability.

Are the roadmap priorities driven more by enterprise document processing needs or by developer/agent pipeline needs?
